### PR TITLE
Restore improvement: index

### LINKED ISSUE #3981: Create batches from all manifest at once
This has the following benefits:
- increasing file pool allows for creating more evenly distributed batches
- removes the current waiting barrier when SM waited for one manifest to be fully restored before restoring another one
- changes the workflow so that it is easier to restore multiple tables in parallel from the same node

Note that files included in a single batch will still originate from a single manifest, but different batches restored in parallel by different nodes could come from different manifests.

### PR DESCRIPTION
**Pull request by Michal-Leszczynski**
This PR is the core of restore improvements prepared for SM 3.4.
It refactors the general workflow from restoring location by location, manifest by manifest, table by table, to a version where nothing waits on nothing! All of the remote sstable dirs are indexed at the beginning, fed to batch dispatcher, which controls batch creation. For now, the batch dispatcher creates batches in the same way as before, but it will be imporved as a part of #3979.

Rebased on #4040.

Fixes #3981

### CODE REVIEW

**Comment by mykaul** [comment]
> This PR is the core of restore improvements prepared for SM 3.4. It refactors the general workflow from restoring location by location, manifest by manifest, table by table, to a version where nothing waits on nothing! All of the remote sstable dirs are indexed at the beginning, fed to batch dispatcher, which controls batch creation. For now, the batch dispatcher creates batches in the same way as before, but it will be imporved as a part of #3979.
> 
> Rebased on #4040.
> 
> Fixes #3981

General note - I'm lacking some documentation here. For example, how do we sort them (if we do) after indexing? do we have any way to control it, etc.

**Comment by karol-kokoszka** [comment]
We have just the document (design doc?) describing the restore in Scylla Manager 3.1.
https://docs.google.com/document/d/1ugf6XCs28c4pwKsw9Lbu_EFb6ZSgMpJ_-Usf-Q9XAP4/edit#heading=h.f7jv2tbf2ofd

I agree that all the details that are going to be included into 3.4 must be documented.

We can either make it public, by editing README.md of https://github.com/scylladb/scylla-manager/tree/master/pkg/service/restore and putting reference to it in main README, or by just preparing new design doc documenting all the changes/updates or current state.

The design, list of changes were introduced on our meeting where we wanted to understand the bottlenecks from the current implementation. We haven't created formal design doc.... but just set of issues.

Issue for the documentation https://github.com/scylladb/scylla-manager/issues/4051

File: pkg/service/restore/index.go [hunk 1]
```diff
@@ -71,6 +72,12 @@ func (w *tablesWorker) indexLocationWorkload(ctx context.Context, location Locat
 	if err != nil {
 		return LocationWorkload{}, errors.Wrap(err, "create remote dir workloads")
 	}
+	if w.target.Continue {
+		rawWorkload, err = w.filterPreviouslyRestoredSStables(rawWorkload)
```
Comment (karol-kokoszka) [inline pkg/service/restore/index.go:77]: Maybe worth to log the rawWorkload ?
Without including name of every SSTable, but by just counting how many of SSTables are per RemoteDirWorkload and how much of the data ?

File: pkg/service/restore/index.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+// SSTable represents files creating a single sstable.
+type SSTable struct {
+	ID    string
+	Files []string
+}
+
+// IndexWorkload returns sstables to be restored aggregated by location, table and remote sstable dir.
+func (w *tablesWorker) IndexWorkload(ctx context.Context, locations []Location) ([]LocationWorkload, error) {
```
Comment (karol-kokoszka) [inline pkg/service/restore/index.go:59]: Could it be logged somehow ? I mean the LocationWorkload without including name and size of every SSTable, but just size of all SSTables in the current chunk + amount of SSTables.
Would it help somehow with debugging ?

**Review by karol-kokoszka — APPROVED** [review]
:+1:

**Comment by Michal-Leszczynski** [comment]
@karol-kokoszka I addressed all the comments, please take a look.

**Comment by Michal-Leszczynski** [comment]
Failures are known issues/flakes:
- https://github.com/scylladb/scylla-manager/issues/4050
- https://github.com/scylladb/scylla-manager/issues/3848

File: pkg/service/restore/index.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+	if err != nil {
+		return LocationWorkload{}, errors.Wrap(err, "create remote dir workloads")
+	}
+	return aggregateLocationWorkload(rawWorkload), nil
+}
+
+func (w *tablesWorker) createRemoteDirWorkloads(ctx context.Context, location Location) ([]RemoteDirWorkload, error) {
+	var rawWorkload []RemoteDirWorkload
+	err := w.forEachManifest(ctx, location, func(m ManifestInfoWithContent) error {
+		return m.ForEachIndexIterWithError(nil, func(fm FilesMeta) error {
```
[the remark this answers is not in the document]
Comment (karol-kokoszka) [reply pkg/service/restore/index.go:90]: It's a bit of misleading. Nil is basically empty slice, but let's accept it now.

**Review by karol-kokoszka — APPROVED** [review]

**Comment by karol-kokoszka** [comment]
Please check why Sanity check fails.

**Comment by Michal-Leszczynski** [comment]
>> Failures are known issues/flakes:
>> - https://github.com/scylladb/scylla-manager/issues/4050
>> - https://github.com/scylladb/scylla-manager/issues/3848

> Please check why Sanity check fails.

Because of the known flake: https://github.com/scylladb/scylla-manager/issues/3848

**Pull request merged**